File: test/jskills/elo/DuellingEloTest.java
using Moserware.Skills;
using Moserware.Skills.Elo;
using NUnit.Framework;

namespace UnitTests.Elo
{
    [TestFixture]
    public class DuellingEloTest
    {
        private final double ErrorTolerance = 0.1;

        [Test]
        public void TwoOnTwoDuellingTest()
        {
            calculator = new DuellingEloCalculator(new GaussianEloCalculator());

            player1 = new Player(1);
            player2 = new Player(2);

            gameInfo = GameInfo.DefaultGameInfo;

            team1 = new Team()
                .AddPlayer(player1, gameInfo.DefaultRating)
                .AddPlayer(player2, gameInfo.DefaultRating);

            player3 = new Player(3);
            player4 = new Player(4);

            team2 = new Team()
                        .AddPlayer(player3, gameInfo.DefaultRating)
                        .AddPlayer(player4, gameInfo.DefaultRating);

            teams = Teams.Concat(team1, team2);
            newRatingsWinLose = calculator.CalculateNewRatings(gameInfo, teams, 1, 2);

            // TODO: Verify?
            AssertRating(37, newRatingsWinLose[player1]);
            AssertRating(37, newRatingsWinLose[player2]);
            AssertRating(13, newRatingsWinLose[player3]);
            AssertRating(13, newRatingsWinLose[player4]);

            quality = calculator.CalculateMatchQuality(gameInfo, teams);
            assertEquals(1.0, quality, 0.001);
        }

        private static void AssertRating(double expected, Rating actual)
        {
            assertEquals(expected, actual.Mean, ErrorTolerance);
        }
    }
}

File: test/jskills/elo/DuellingEloTest.scala
package jskills.elo

import java.util.Arrays
import java.util.Collection
import java.util.Map

import jskills.GameInfo
import jskills.IPlayer
import jskills.ITeam
import jskills.Player
import jskills.Rating
import jskills.SkillCalculator
import jskills.Team

import org.junit.Test
import org.junit.Assert._

class DuellingEloTest {
  val ErrorTolerance = 0.1

  @Test
  def twoOnTwoDuellingTest() {
    val calculator = new DuellingEloCalculator(new GaussianEloCalculator())

    val gameInfo = GameInfo.getDefaultGameInfo()

    val player1 = new Player(1)
    val player2 = new Player(2)

    val team1 = new Team()
      .addPlayer(player1, gameInfo.getDefaultRating())
      .addPlayer(player2, gameInfo.getDefaultRating())

    val player3 = new Player(3)
    val player4 = new Player(4)

    val team2 = new Team()
      .addPlayer(player3, gameInfo.getDefaultRating())
      .addPlayer(player4, gameInfo.getDefaultRating())

    val teams = Arrays.asList(team1, team2)
    val newRatingsWinLose = calculator.calculateNewRatings(gameInfo, teams, Seq(2, 1))

    // TODO: Verify?
    AssertRating(37, newRatingsWinLose.get(player1))
    AssertRating(37, newRatingsWinLose.get(player2))
    AssertRating(13, newRatingsWinLose.get(player3))
    AssertRating(13, newRatingsWinLose.get(player4))

    val quality = calculator.calculateMatchQuality(gameInfo, teams)
    assertEquals(1.0, quality, 0.001)
  }

  def AssertRating(expected: Double, actual: Rating) {
    assertEquals(expected, actual.getMean(), ErrorTolerance)
  }
}